integrated_channels/canvas/client.py

```diff
@@ -173,18 +174,25 @@ def _create_session(self):
         """
         Instantiate a new session object for use in connecting with Canvas. Each enterprise customer
         connecting to Canvas should have a single client session.
+        Will only create a new session if token expiry has been reached
         """
-        if self.session:
-            self.session.close()
-        # Create a new session with a valid token
-        oauth_access_token = self._get_oauth_access_token(
-            self.enterprise_configuration.client_id,
-            self.enterprise_configuration.client_secret,
-        )
-        session = requests.Session()
-        session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
-        session.headers['content-type'] = 'application/json'
-        self.session = session
+        now = datetime.utcnow()
+        if self.session is None or self.expires_at is None or now >= self.expires_at:
+            # need new session if session expired, or not initialized
+            if self.session:
+                self.session.close()
+            # Create a new session with a valid token
+            oauth_access_token, expires_in = self._get_oauth_access_token(
+                self.enterprise_configuration.client_id,
+                self.enterprise_configuration.client_secret,
+            )
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['content-type'] = 'application/json'
+            self.session = session
+            # expiry expected after `expires_in` seconds
+            if expires_in is not None:
+                self.expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
```

Comment:
I didn't look into the expiration handling from Canvas, is it always in seconds? What is the actual amount of time until we lose the token?

Reply:
yup, always in seconds (1 hour by default)

Reply:
https://canvas.instructure.com/doc/api/file.oauth.html#using-refresh-tokens

Reply:
regardless, this will use the expires_in (Seconds) as a way to setup when we should expect expiry